core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1698,150 +1269,348 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
+class LogDirEventNotificationHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChildChangeHandler {
+  override val path: String = LogDirEventNotificationZNode.path
+
+  override def handleChildChange: Unit = eventManager.put(controller.LogDirEventNotification)
 }
 
-/**
- * Starts the partition reassignment process unless -
- * 1. Partition previously existed
- * 2. New replicas are the same as existing replicas
- * 3. Any replica in the new set of replicas are dead
- * If any of the above conditions are satisfied, it logs an error and removes the partition from list of reassigned
- * partitions.
- */
-class PartitionReassignmentListener(controller: KafkaController, eventManager: ControllerEventManager) extends IZkDataListener with Logging {
-  override def handleDataChange(dataPath: String, data: Any): Unit = {
-    val partitionReassignment = ZkUtils.parsePartitionReassignmentData(data.toString)
-    eventManager.put(controller.PartitionReassignment(partitionReassignment))
-  }
+class PartitionModificationsHandler(controller: KafkaController, eventManager: ControllerEventManager, topic: String) extends ZNodeChangeHandler {
+  override val path: String = TopicZNode.path(topic)
 
-  override def handleDataDeleted(dataPath: String): Unit = {}
+  override def handleCreation: Unit = ???
+
+  override def handleDeletion: Unit = ???
+
+  override def handleDataChange: Unit = eventManager.put(controller.PartitionModifications(topic))
 }
 
-class PartitionReassignmentIsrChangeListener(controller: KafkaController, eventManager: ControllerEventManager,
-                                             topic: String, partition: Int, reassignedReplicas: Set[Int]) extends IZkDataListener with Logging {
-  override def handleDataChange(dataPath: String, data: Any): Unit = {
-    eventManager.put(controller.PartitionReassignmentIsrChange(TopicAndPartition(topic, partition), reassignedReplicas))
-  }
+class TopicDeletionHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChildChangeHandler {
+  override val path: String = DeleteTopicsZNode.path
 
-  override def handleDataDeleted(dataPath: String): Unit = {}
+  override def handleChildChange: Unit = eventManager.put(controller.TopicDeletion)
 }
 
-/**
- * Called when replica leader initiates isr change
- */
-class IsrChangeNotificationListener(controller: KafkaController, eventManager: ControllerEventManager) extends IZkChildListener with Logging {
-  override def handleChildChange(parentPath: String, currentChilds: java.util.List[String]): Unit = {
-    import JavaConverters._
-    eventManager.put(controller.IsrChangeNotification(currentChilds.asScala))
-  }
+class PartitionReassignmentHandler(controller: KafkaController, eventManager: ControllerEventManager) extends ZNodeChangeHandler {
+  override val path: String = ReassignPartitionsZNode.path
+
+  override def handleCreation: Unit = eventManager.put(controller.PartitionReassignment)
+
```

Comment:
This will be called when the partition reassignment completes. So, we probably want to make this a no-op instead of throwing an exception.

Reply:
That's a good point. See my above comment describing several ways we can address this.